--- src/sqliteLimit.h
+++ src/sqliteLimit.h
@@ -102,25 +102,26 @@
 ** the main database table and for temporary tables.
 **
 ** IMPLEMENTATION-OF: R-30185-15359 The default suggested cache size is -2000,
 ** which means the cache size is limited to 2048000 bytes of memory.
 ** IMPLEMENTATION-OF: R-48205-43578 The default suggested cache size can be
 ** altered using the SQLITE_DEFAULT_CACHE_SIZE compile-time options.
 */
 #ifndef SQLITE_DEFAULT_CACHE_SIZE
-# define SQLITE_DEFAULT_CACHE_SIZE  -2000
+# define SQLITE_DEFAULT_CACHE_SIZE   5000
 #endif
 
 /*
 ** The default number of frames to accumulate in the log file before
 ** checkpointing the database in WAL mode.
 */
 #ifndef SQLITE_DEFAULT_WAL_AUTOCHECKPOINT
-# define SQLITE_DEFAULT_WAL_AUTOCHECKPOINT  1000
+# define SQLITE_DEFAULT_WAL_AUTOCHECKPOINT \
+        SQLITE_DEFAULT_JOURNAL_SIZE_LIMIT / SQLITE_DEFAULT_PAGE_SIZE
 #endif
 
 /*
 ** The maximum number of attached databases.  This must be between 0
 ** and 125.  The upper bound of 125 is because the attached databases are
 ** counted using a signed 8-bit integer which has a maximum value of 127
 ** and we have to allow 2 extra counts for the "main" and "temp" databases.
 */
