--- Makefile.in
+++ Makefile.in
@@ -1084,17 +1084,17 @@ FTS5_SRC = \
    $(TOP)/ext/fts5/fts5_index.c \
    fts5parse.c fts5parse.h \
    $(TOP)/ext/fts5/fts5_storage.c \
    $(TOP)/ext/fts5/fts5_tokenize.c \
    $(TOP)/ext/fts5/fts5_unicode2.c \
    $(TOP)/ext/fts5/fts5_varint.c \
    $(TOP)/ext/fts5/fts5_vocab.c  \
 
-fts5parse.c:	$(TOP)/ext/fts5/fts5parse.y lemon 
+fts5parse.c:	$(TOP)/ext/fts5/fts5parse.y lemon$(BEXE)
 	cp $(TOP)/ext/fts5/fts5parse.y .
 	rm -f fts5parse.h
 	./lemon$(BEXE) $(OPTS) fts5parse.y
 
 fts5parse.h: fts5parse.c
 
 fts5.c: $(FTS5_SRC)
 	$(TCLSH_CMD) $(TOP)/ext/fts5/tool/mkfts5c.tcl
